mirgecom/diffusion.py

```diff
@@ -31,26 +31,48 @@
 """
 
 import abc
-import math
 import numpy as np
 import numpy.linalg as la  # noqa
 from pytools.obj_array import make_obj_array, obj_array_vectorize_n_args
 from meshmode.mesh import BTAG_ALL, BTAG_NONE  # noqa
 from meshmode.dof_array import thaw
+from grudge.symbolic.primitives import DOFDesc, QTAG_NONE
 from grudge.eager import interior_trace_pair, cross_rank_trace_pairs
-from grudge.symbolic.primitives import TracePair
+from grudge.symbolic.primitives import TracePair, as_dofdesc
 
 
-def _q_flux(discr, alpha, u_tpair):
-    normal = thaw(u_tpair.int.array_context, discr.normal(u_tpair.dd))
-    flux_weak = math.sqrt(alpha)*u_tpair.avg*normal
-    return discr.project(u_tpair.dd, "all_faces", flux_weak)
+def _q_flux(discr, var_diff_quad_tag, alpha, u_tpair):
+    actx = u_tpair.int.array_context
 
+    dd = u_tpair.dd
+    dd_quad = dd.with_qtag(var_diff_quad_tag)
+    dd_allfaces_quad = dd_quad.with_dtag("all_faces")
 
-def _u_flux(discr, alpha, q_tpair):
-    normal = thaw(q_tpair.int[0].array_context, discr.normal(q_tpair.dd))
-    flux_weak = math.sqrt(alpha)*np.dot(q_tpair.avg, normal)
-    return discr.project(q_tpair.dd, "all_faces", flux_weak)
+    normal = thaw(actx, discr.normal(dd))
+
+    alpha_quad = discr.project("vol", dd_quad, alpha)
+    sqrt_alpha_quad = actx.np.sqrt(alpha_quad)
+
+    flux_quad = discr.project(dd, dd_quad, -u_tpair.avg*normal)
```

Comment:
Interesting. How come we don't do that for Euler, etc.? Is that on the to-do list for later?

Reply:
I think I misunderstood this comment when I read it yesterday. The multiplication by the normal doesn't demand overintegration in general, it's just that since we're already doing it we want to use the more accurate normal from the quadrature discretization. Something like that?

Reply:
*If* we use curvilinear geometry or quad/hex elements, *then* the normal (as a function on the element) won't be a constant. As a result, the multiplication has the potential to drive an aliasing instability, and overintegration is generically necessary.

Does that explain it better?

Reply:
I think so... Does that mean that in the curvilinear case the other operators (Euler/wave/whatever) will likely need overintegration for the normal as well?

Reply:
Yep. *If* we end up doing curvi geometry.